Extensions.md

```diff
@@ -0,0 +1,127 @@
+3. consider its final form where the long form header for `32767: binary` may not be present.
+
+Once the design is ratified the new `FileMetaData` encoding is made final with the following migration plan. For the next N years writers will write both the Thrift and the flatbuffer `FileMetaData`. It will look much like its private form except the flatbuffer IDL may be different:
+
+    N-1 bytes | Thrift compact protocol encoded FileMetadata (minus \0 thrift stop field)
+    4 bytes   | 08 FF FF 01 (long form header for 32767: binary)
+    1-5 bytes | ULEB128(K+28) encoded size of the extension
+    K bytes   | Flatbuffers representation (v1) of FileMetaData
+    4 bytes   | little-endian crc32(flatbuffer)
+    4 bytes   | little-endian size(flatbuffer)
+    4 bytes   | little-endian crc32(size(flatbuffer))
+    16 bytes  | some-other-UUID
+    1 byte    | \0 (thrift stop field)
+    4 bytes   | PAR1
+
+After the migration period, the end of the Parquet file may look like this:
+
+    K bytes   | Flatbuffers representation (v1) of FileMetaData
+    4 bytes   | little-endian crc32(flatbuffer)
+    4 bytes   | little-endian size(flatbuffer)
+    4 bytes   | little-endian crc32(size(flatbuffer))
+    4 bytes   | PAR3
+
+In this example, we see several design decisions for the extension at play:
+
+* There is a new UUID2 for the accepted change to the standard and now the Thrift `FileMetaData` cannot be extended itself.  
+* The length of the footer and the crc32 of the length itself, guarantees that new readers will not overshoot reading bytes in case of corrupt bits in these critical 8 bytes of the file.  
+* The crc32 of the flatbuffer representation enhances Parquet to have crc32 for metadata as well which is arguably more important than crc32 for data.  
+* The new encoding itself, which MUST contain some way to be extended in the future (much like Thrift does with this specification).
+
+### Encoding
+
+The community experiments with a new encoding extension. At the same time they want to keep the newly encoded Parquet files open for everyone to read. So they add a new encoding via an extension to the ColumnMetaData struct. The extension stores offsets in the Parquet file where the new and duplicate encoded data for this column lives. The new writer carefully places all the new encodings at the start of the row group and all the old encodings at the end of the row group. This layout minimizes disruption for readers unaware of the new encodings.
+
+In its private form Parquet files look like so:
+
+    4 bytes   | PAR1
+              |             | Column b (new encoding)
+              |             | Column c (new encoding)
+    R bytes   |  Row Group  | Column a
+              |     0       | Column d
+              |             | Column b (old encoding)
+              |             | Column c (old encoding)
+    4 bytes   | PAR1
```

Comment:
```suggestion
    4 bytes   | PAR1
    ----------+-------------+----------------------------
              |             | Column b (new encoding)
              |             | Column c (new encoding)
    R bytes   |  Row Group  | Column a
              |     0       | Column d
              |             | Column b (old encoding)
              |             | Column c (old encoding)
    ----------+-------------+----------------------------
    F bytes   |FileMetaData |
    ----------+-------------+----------------------------
    4 bytes   | PAR1
```

It looks a little bit confusing to me if we omit the `FileMetaData` section. Propose the change for completion with a little stylish modification.

Reply:
I committed it because the extension in this case is inside `ColumnMetaData` and it would be one enum + offset/len. If we add this here perhaps I should flesh out that part as well?

Reply:
My confusion comes from the fact that it is actually the file layout. How about adding an example on what the extension looks like in the `ColumnMetaData`?

Reply:
Added. Let me know if it helps.